### PR TITLE
Feature/inventory refused

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

File: inc/inventory/request.class.php [hunk 1]
```diff
@@ -444,13 +434,34 @@ public function inError() {
...
       ];
+
+      if (count($items) == 1) {
+         $item = $items[0];
+         $status += [
+            'itemtype' => $item->getType(),
+            'items_id' => $item->fields['id']
+         ];
+      } else if (count($items)) {
+         $itemtype = null;
```
Comment (cedric-anne) [inline inc/inventory/request.class.php:450]: A comment here can be usefull
```suggestion
         // Defines 'itemtype' only if all items has same type
         $itemtype = null;
```

File: inc/refusedequipment.class.php [hunk 2]
```diff
@@ -238,4 +240,38 @@ public function isDynamic() {
...
+   /**
+    * Handle inventory request, and returns redirection url
+    *
+    * @return string
+    */
+   public function handleInventoryRequest(Request $request) {
+      $status = $request->getInventoryStatus();
+
+      if (!isset($status['itemtype'])) {
+         //no itemtype found, redirect to refused logs - should not happen
```
Comment (cedric-anne) [inline inc/refusedequipment.class.php:253]: Maybe a message should be added. Without a message, user will not understand why nothing is done.

File: inc/refusedequipment.class.php [hunk 3]
```diff
@@ -238,4 +240,38 @@ public function isDynamic() {
...
+         return $this->getSearchURL();
+      } else if ($status['itemtype'] === RefusedEquipment::class) {
+         Session::addMessageAfterRedirect(
+            __('Inventory is still refused')
+         );
+         return $this->getSearchURL();
+      } else {
+         $this->delete(['id' => $this->fields['id']], true);
+         Session::addMessageAfterRedirect(
+            __('Inventory is successufl, refused entry log has been removed')
```
Comment (cedric-anne) [inline inc/refusedequipment.class.php:263]:
```suggestion
            __('Inventory is successful, refused entry log has been removed.')
```

File: inc/refusedequipment.class.php [hunk 4]
```diff
@@ -238,4 +240,38 @@ public function isDynamic() {
...
+    */
+   public function handleInventoryRequest(Request $request) {
+      $status = $request->getInventoryStatus();
+
+      if (!isset($status['itemtype'])) {
+         //no itemtype found, redirect to refused logs - should not happen
+         return $this->getSearchURL();
+      } else if ($status['itemtype'] === RefusedEquipment::class) {
+         Session::addMessageAfterRedirect(
+            __('Inventory is still refused')
```
Comment (cedric-anne) [inline inc/refusedequipment.class.php:257]:
```suggestion
            __('Inventory is still refused.')
```

File: tests/functionnal/Glpi/Inventory/Inventory.php [hunk 5]
```diff
@@ -3176,6 +3211,46 @@ public function testImportRefusedFromAssetRules() {
...
+      );
+
+      //test inventory, will be accepted
+      /*if ($inventory->inError()) {
+         foreach ($inventory->getErrors() as $error) {
+            var_dump($error);
+         }
+      }
+      $this->boolean($inventory->inError())->isFalse();
+      $this->array($inventory->getErrors())->isEmpty();*/
```
Comment (cedric-anne) [inline tests/functionnal/Glpi/Inventory/Inventory.php:3239]:
```suggestion
```

File: inc/refusedequipment.class.php (hunk 2 above)
Comment (trasher) [reply inc/refusedequipment.class.php]: I've dropped this part, this should not happen

File: tests/units/Glpi/Inventory/Request.php [hunk 6]
```diff
@@ -57,32 +57,25 @@ function () use ($request) {
...
-      )
-         ->isInstanceOf('\RuntimeException')
-         ->hasMessage('Unknown mode 42');
    }
 
    public function testProlog() {
       $data = "<?xml version=\"1.0\"?>\n<REQUEST><DEVICEID>atoumized-device</DEVICEID><QUERY>PROLOG</QUERY></REQUEST>";
       $request = new \Glpi\Inventory\Request;
-      $request->setCompression('application/xml');
+      $request->HandleContentType('application/xml');
```
Comment (cedric-anne) [inline tests/units/Glpi/Inventory/Request.php:78]:
```suggestion
      $request->handleContentType('application/xml');
```

File: tests/units/Glpi/Inventory/Request.php [hunk 7]
```diff
@@ -105,7 +98,7 @@ public function testSnmpQuery($query) {
       $request = new \mock\Glpi\Inventory\Request();
       $this->calling($request)->inventory = null;
       $this->calling($request)->prolog = null;
-      $request->setCompression('Application/xml');
+      $request->HandleContentType('Application/xml');
```
Comment (cedric-anne) [inline tests/units/Glpi/Inventory/Request.php:101]:
```suggestion
      $request->handleContentType('Application/xml');
```

File: tests/units/Glpi/Inventory/Request.php [hunk 8]
```diff
@@ -114,7 +107,7 @@ public function testSnmpQuery($query) {
    public function testWrongQuery() {
       $data = "<?xml version=\"1.0\"?>\n<REQUEST><DEVICEID>atoumized-device</DEVICEID><QUERY>UNKNWON</QUERY></REQUEST>";
       $request = new \Glpi\Inventory\Request;
-      $request->setCompression('application/xml');
+      $request->HandleContentType('application/xml');
```
Comment (cedric-anne) [inline tests/units/Glpi/Inventory/Request.php:110]:
```suggestion
      $request->handleContentType('application/xml');
```

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**